### ISSUE TITLE
Proposal 3: don't only store problems

### ISSUE THREAD

**Issue by macinspak** [Anamico/node-red-contrib-alarm#27]
Originally from #24, submitted by @bartbutenaers

I have not analyzed the internal kitchen of these nodes yet into detail, so perhaps this is already possible somehow.  

When I want to switch my alarm from Off to Away, I want to show a list of the current problems.  In the following demo my Node-RED flow sees that there is still one door open, which it shows to me:

![demo_popup](https://user-images.githubusercontent.com/14224149/128591892-5fde1d92-fec1-4ae5-b6ff-d8b4b9a1b584.gif)

Then I can decide whether I cancel the state-change and close the door first.  Or I can continue switching the state.

To keep my flow simple, it would be nice if the alarm-nodes could store this information, and make it queryable.  Indeed the sensor node gets all values from all sensors, and it knows (based on the last sensor value and the Trigger dropdown options) which sensors currently have detected an issue.

**Comment by macinspak** [comment]
Yes, I had also been thinking about the best way to implement this.

I guess again I had been postponing it myself as I tend to arm the house after leaving and not had this issue. But it is a standard sort of function. Perhaps having real-time nodes always outputting "open or close" state will allow people to build their own led dashboard equivalent, and maybe some logic or additional parameters that will auto-fail setting a mode and returning a separate error output connection if the alarm would immediately fire? Then standard behaviour remains, but you can set an "exit delay" for any mode via that node configuration or passing in a parameter as part of the msg. Then you could fail to switch the mode and push a push notification or "pip" an alarm or something? Or play a sound on your alexa or homepod?

What do people think?

**Comment by bartbutenaers** [comment]
> I had been postponing it myself as I tend to arm the house after leaving and not had this issue

Indeed when you activate remotely your alarm (via e.g. a smart phone), this is not really useful.  Because when you already left your house, you cannot do this:
1. Press the "Cancel" button
2. Close the front door (which was still open as mentioned in the popup)
3. Activate the Away mode again (and now no popup anymore since all doors are closed.

Although it is still a bit useful, because you can blame your wife and kids in the car that they again forgot to close some doors ;-)

Currently I do it like this in my node:
1. When a sensor value arrives, I calculate whether it is a trigger value or not (similar like you do it in your Sensor node).
2. Then I store per sensor (e.g. per topic) the boolean result of that calculation.  So I remember whether the last value was a trigger value or not.
3. When I want to activate my alarm, I ask my node a list of all sensors (e.g. topics) whose last value was a trigger value.
4. I show those sensors in a popup.

In my case this is all inside a single node, but I like your approach of multiple nodes a LOT more!  Because I have a lot of wiring towards the same node, which again leads to a spaghetti flow. 
To avoid ruining your nice concept, could it perhaps be useful to add a new node to your repository?  E.g. a Query node, whose config screen contains a dropdown where you can select a query.  In this case the "Sensors with trigger value" query.  Then we could put that Query node in between the Dashboard nodes in my use case:

![image](https://user-images.githubusercontent.com/14224149/128622554-bc7a4068-c415-4c7c-b51d-d1a14dd42393.png)

Note that the issues popup need only to be showed if there are any sensors whose last value was a trigger value!  Because it has no use to show an empty popup...

Therefore it would be useful that such a node had two outputs: 
+ when the query resultset contains 1 or more entries, then an output message is being send on the first output.
+ when the query resultset contains 0 entries, then an output message is being send on the first output (containing an empty array) **and** also on the second output.

The second output can be omitted, but then again you need to add an extra Switch node to check whether the resultset is empty.  Resulting again in a less clean flow...

But there might be better ways to implement this in this node suite ...

**Comment by macinspak** [comment]
Going to think about this one. I like the idea. We can use it to trigger warning "pips" or a push notification to let you know you armed the house, but the back door is still open so you had better turn around and fix it.

**Comment by macinspak** [comment]
[switch.html](https://github.com/node-red/node-red/blob/master/packages/node_modules/%40node-red/nodes/core/function/10-switch.html)
[switch.js](https://github.com/node-red/node-red/blob/master/packages/node_modules/%40node-red/nodes/core/function/10-switch.js)

**Comment by bartbutenaers** [comment]
Hi @macinspak,
If you should have had time to think about a proposal, it would be nice if we could discuss about it.
Thanks !!
Bart

**Comment by macinspak** [comment]
It has indeed also been bugging me on how best to implement. I mean, if you
look at traditional alrams, they used to do that alarm pip if there was an
unsecured zone, but I cannot remember if they refused to arm. Maybe it is
OK to arm, and then they walk out the front door, because CLOSING the door
does not trigger the already armed system? Maybe as it only should alarm on
OPEN then it's ok. It just means if you arm it and do NOT close the front
door, then that zone or sector is disabled. I think from memory it used to
do that on the old alarm systems right? I mean it would disable the
unsecured zones?

I guess we just need to work out the semantic to make the most sense. What
are the use-cases?
1. woman pushes button and walks out front door, get's notification they
left the laundry door open?

   - they ignore it and it's disabled? Doesn't matter if it opens and
      closes (banging in the wind as it's latched open)?
      - they walk down the side and pull the door shut and now it's armed
      (because it would only trigger on a new "open" event)?

Is it that simple? Or what other cases do we need to cater for?

Love your thinking by the way, great to have someone now to bounce this off
to get the best outcome.
Andrew
